### PR TITLE
WIP: Add Endpoints for Contributions Table

### PR DESCRIPTION
**Pull request by nigelpain**
## What

[Link to story](https://dsdmoj.atlassian.net/jira/software/projects/LCAM/boards/881?selectedIssue=LCAM-375)

Added three new endpoints to the application to perform operations against the contributions table based on those specified in the calc_appeal_contrib stored procedure:
- *find* Take a repId, and uses this to query the contributions table for any current matching rows (that is those where the active column has a value of'Y'). If no row is found then a 404 response is returned.
- *update* Takes the json passed through in the request body and performs validation against it according to DB constraints. It then queries the contributions table to find the matching row to update, if no row is found returns a 404 response. Otherwise maps the updated values from the request body to the entity and saves them to the database.
- *create* Takes json passed through in the request body and performs validation against it according to DB constraints. Then queries the contributions table to see if there is an existing row. If there is it updates this row to be the 'old' row by setting the replaced date, active to 'Y' and latest to 'N'. Then the request data is mapped to an entity before saving it to the database.

## TODO
- Write integration tests for ContributionsController.
- Get PR reviewed and action any comments.

## Checklist

Before you ask people to review this PR:

- [ ] Tests should be passing: `./gradlew test`
- [ ] Github should not be reporting conflicts; you should have recently run `git rebase main`.
- [ ] Avoid mixing whitespace changes with code changes in the same commit. These make diffs harder to read and conflicts more likely.
- [ ] You should have looked at the diff against main and ensured that nothing unexpected is included in your changes.
- [ ] You should have checked that the commit messages say why the change was made.

### CODE REVIEW

File: maat-court-data-api/src/test/java/gov/uk/courtdata/contributions/service/ContributionsServiceTest.java [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+        ContributionsEntity contributionsEntity = ContributionsEntity.builder().build();
+        when(contributionsImpl.find(anyInt())).thenReturn(contributionsEntity);
+        when(contributionsImpl.update(any(ContributionsEntity.class))).thenReturn(contributionsEntity);
+        when(contributionsMapper.mapEntityToDTO(any(ContributionsEntity.class))).thenReturn(ContributionsDTO.builder().id(testId).build());
+
+        ContributionsDTO contributionsDTO = contributionsService.update(UpdateContributions.builder().id(testId).build());
+
+        assertThat(contributionsDTO.getId()).isEqualTo(testId);
+        verify(contributionsImpl).update(contributionsEntity);
+    }
```
Comment (AafAnsari) [inline maat-court-data-api/src/test/java/gov/uk/courtdata/contributions/service/ContributionsServiceTest.java:67]: mock calls for `contributionsImpl.find()` and `contributionsMapper.mapEntityToDTO` can be verified. Same goes for all test cases where mock is used

File: maat-court-data-api/src/main/java/gov/uk/courtdata/repository/ContributionsRepository.java [hunk 2]
```diff
@@ -12,4 +16,14 @@ public interface ContributionsRepository extends JpaRepository<ContributionsEnti
     Integer countAllByRepId(Integer repId);
 
     List<ContributionsEntity> findAllByRepId(Integer repId);
+
+    ContributionsEntity findByRepIdAndLatestIsTrue(int repId);
+
+    @Modifying
+    @Query(value = "UPDATE TOGDATA.CONTRIBUTIONS SET REPLACED_DATE = TRUNC(SYSDATE), ACTIVE = 'N' WHERE REP_ID = :repId AND EFFECTIVE_DATE >= :effDate", nativeQuery = true)
+    void setEntryAsInactive(@Param("repId") Integer repId, @Param("effDate") LocalDate effDate);
+
```
Comment (AafAnsari) [inline maat-court-data-api/src/main/java/gov/uk/courtdata/repository/ContributionsRepository.java:25]: method can be renamed to `updateExistingContributionToInactive`

File: maat-court-data-api/src/main/java/gov/uk/courtdata/repository/ContributionsRepository.java [hunk 3]
```diff
@@ -12,4 +16,14 @@ public interface ContributionsRepository extends JpaRepository<ContributionsEnti
...
+    @Modifying
+    @Query(value = "UPDATE TOGDATA.CONTRIBUTIONS SET LATEST = 'N' WHERE REP_ID = :repId", nativeQuery = true)
+    void setEntryAsPrior(@Param("repId") Integer repId);
 }
```
Comment (AafAnsari) [inline maat-court-data-api/src/main/java/gov/uk/courtdata/repository/ContributionsRepository.java:29]: method can be renamed to `updateExistingContributionToPrior`

File: maat-court-data-api/src/main/java/gov/uk/courtdata/contributions/impl/ContributionsImpl.java [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+    public ContributionsEntity findLatest(int repId) {
+        log.debug("Retrieving latest contributions entry for rep ID {}", repId);
+        return contributionsRepository.findByRepIdAndLatestIsTrue(repId);
+    }
+
+    public ContributionsEntity update(ContributionsEntity contributionsEntity) {
+        return contributionsRepository.saveAndFlush(contributionsEntity);
+    }
+
+    public Optional<Void> updateInactiveAndPrior(Integer repId, LocalDate effectiveDate) {
```
Comment (AafAnsari) [inline maat-court-data-api/src/main/java/gov/uk/courtdata/contributions/impl/ContributionsImpl.java:35]: method can be renamed to `updateExistingContributions`

**Comment by AafAnsari** [comment]
- `contributions` package is not required as we already have `contribution` package.
- `ContributionsImpl` is not required and the repository calls can be made from the service class itself
- Test methods to follow given_when_then format for all test cases
- For all tests it’s best to use the actual values instead of generic any() where the test is relying on specific value and not any().

**Review by AafAnsari — CHANGES_REQUESTED** [review]

File: maat-court-data-api/src/test/java/gov/uk/courtdata/integration/contributions/ContributionsControllerIntegrationTest.java [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+    @Autowired
+    MockMvc mvc;
+    @Autowired
+    private RepOrderRepository repOrderRepository;
+
+    private ContributionsEntity contributionsEntity;
+
+    @BeforeEach
+    public void setUP(@Autowired RepOrderRepository repOrderRepository,
+                      @Autowired ContributionsRepository contributionsRepository) {
```
Comment (AafAnsari) [inline maat-court-data-api/src/test/java/gov/uk/courtdata/integration/contributions/ContributionsControllerIntegrationTest.java:55]: `setUP` typo

File: maat-court-data-api/src/test/java/gov/uk/courtdata/integration/contributions/ContributionsControllerIntegrationTest.java [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+        UpdateContributions updateContributions = TestModelDataBuilder.getUpdateContributions(INVALID_CONTRIBUTION_ID);
+        assertTrue(runBadRequestErrorScenario("Contributions ID: " + INVALID_CONTRIBUTION_ID + " is invalid.",
+                put(ENDPOINT_URL).contentType(MediaType.APPLICATION_JSON)
+                        .content(objectMapper.writeValueAsString(updateContributions))));
+    }
+
+    @Test
+    void givenAValidContent_whenUpdateIsInvoked_theCorrectResponseIsReturned() throws Exception {
+
+
```
Comment (AafAnsari) [inline maat-court-data-api/src/test/java/gov/uk/courtdata/integration/contributions/ContributionsControllerIntegrationTest.java:111]: newline can be removed

**Review by AafAnsari — CHANGES_REQUESTED** [review]

File: maat-court-data-api/src/main/java/gov/uk/courtdata/contributions/controller/ContributionsController.java [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+                    schema = @Schema(implementation = ErrorDTO.class)
+            )
+    )
+    @ApiResponse(responseCode = "500",
+            description = "Internal server error",
+            content = @Content(mediaType = MediaType.APPLICATION_JSON_VALUE,
+                    schema = @Schema(implementation = ErrorDTO.class)
+            )
+    )
+    public ResponseEntity<ContributionsDTO> update(@Valid @RequestBody UpdateContributions updateContributions) {
```
[the remark this answers is not in the document]
Comment (loganathanMOJ) [reply maat-court-data-api/src/main/java/gov/uk/courtdata/contributions/controller/ContributionsController.java:88]: This need for schema validation (eg @NotNull)

File: maat-court-data-api/src/test/java/gov/uk/courtdata/contributions/impl/ContributionsImplTest.java [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+
+import static org.assertj.core.api.AssertionsForClassTypes.assertThat;
+import static org.mockito.Mockito.*;
+
+
+@ExtendWith(MockitoExtension.class)
+class ContributionsImplTest {
+
+    private static final Integer TEST_REP_ID = 999;
+
```
[the remark this answers is not in the document]
Comment (loganathanMOJ) [reply maat-court-data-api/src/test/java/gov/uk/courtdata/contributions/impl/ContributionsImplTest.java]: Fixed now

File: maat-court-data-api/src/test/java/gov/uk/courtdata/integration/contributions/ContributionsControllerIntegrationTest.java (hunk 5 above)
Comment (loganathanMOJ) [reply maat-court-data-api/src/test/java/gov/uk/courtdata/integration/contributions/ContributionsControllerIntegrationTest.java:55]: Thanks. fixed now

File: maat-court-data-api/src/main/java/gov/uk/courtdata/model/contributions/Contributions.java [hunk 9]
```diff
@@ -0,0 +1,50 @@
+package gov.uk.courtdata.model.contributions;
```
Comment (stone-dj) [inline maat-court-data-api/src/main/java/gov/uk/courtdata/model/contributions/Contributions.java:1]: I understand this is still WIP, I'm not sure if the package structure is quite right for some of the classes. Specifically I would expect classes which are only used within a specific functional domain to sit inside a sub-package of that same domain.

i.e. If the Contributions model class is only used by classes in the package `gov.uk.courtdata.contribution` (and it's sub-packages), I would expect the `Contributions` class to sit in the package  `gov.uk.courtdata.contribution.model`

This is consistent with the majority of the existing code and is IMHO a good structure as it allows us to more easily determine the scope of usages of a given class. Also it should reduce package-tangle and ensure dependancies flow vertically in a single direction (minimising inter-package dependancies).

Am more than happy to have a discussion on this and go into more detail when you're next online and take on board any downsides to this approach I may have missed.

https://stackoverflow.com/questions/15321702/what-does-package-tangle-index-data-indicate-in-sonar

**Review by stone-dj — CHANGES_REQUESTED** [review]

**Review by AafAnsari — APPROVED** [review]

**Comment by stone-dj** [comment]
There does not appear to be an agreed upon and documented standard package structure, this is to be discussed and decided on outside of this Pull Request.

**Comment by stone-dj** [comment]
There does not appear to be an agreed upon and documented standard package structure, this is to be discussed and decided on outside of this Pull Request.

**Comment by stone-dj** [comment]
There does not appear to be an agreed upon and documented standard package structure, this is to be discussed and decided on outside of this Pull Request.

**Pull request merged**